### PR TITLE
Fixed #26327 -- Added JsonAgg to contrib.postgres.

### PR DESCRIPTION
**Pull request by atombrella**
Aggregation over multiple columns is not supported (as discussed in the Trac ticket).
https://code.djangoproject.com/ticket/26327

### CODE REVIEW

**Comment by atombrella** [comment]
buildbot, add to whitelist.

**Comment by timgraham** [comment]
Please try to fix the test failures on non-PostgreSQL builds. They should be able to run without installing psycopg2.

**Comment by atombrella** [comment]
It's difficult for me to figure out what causes the import of ``psycopg2``. I experimented a bit with the order of the imports in the added code, but no luck.

**Comment by timgraham** [comment]
Here's the traceback: 
```
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/tests/postgres_tests/test_aggregates.py", line 3, in <module>
    from django.contrib.postgres.aggregates import (
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/aggregates/__init__.py", line 1, in <module>
    from .general import *  # NOQA
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/aggregates/general.py", line 1, in <module>
    from django.contrib.postgres.fields import JSONField
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/fields/__init__.py", line 1, in <module>
    from .array import *  # NOQA
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/fields/array.py", line 4, in <module>
    from django.contrib.postgres.forms import SimpleArrayField
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/forms/__init__.py", line 4, in <module>
    from .ranges import *  # NOQA
  File "/home/jenkins/workspace/pull-requests-trusty/database/spatialite/label/trusty-pr/python/python3.5/django/contrib/postgres/forms/ranges.py", line 1, in <module>
    from psycopg2.extras import DateRange, DateTimeTZRange, NumericRange
```
You might need to move the aggregate imports in a try/except similar to what's in `test_ranges.py`.

**Comment by atombrella** [comment]
OK. It seems a bit ugly to do all this, but it makes the tests run :) I'm not sure which is more appropriate to use, ``object`` or ``None``.

    try: 
        from psycopg2 import X
    except:
        X = object # or None

**Comment by timgraham** [comment]
Sorry, I edited my previous comment to add "in `test_aggregates.py`" but maybe you didn't see that. The workarounds should be in test code, not in `contrib.posgres`.

File: tests/postgres_tests/models.py [hunk 1]
```diff
@@ -149,9 +149,9 @@ def __init__(self, *args, **kwargs):
         super(ArrayFieldSubclass, self).__init__(models.IntegerField())
 
 
-class AggregateTestModel(models.Model):
+class AggregateTestModel(PostgreSQLModel):
```
Comment (timgraham) [inline tests/postgres_tests/models.py]: I can't see why this change is needed?

File: tests/postgres_tests/models.py [hunk 2]
```diff
@@ -149,9 +149,9 @@ def __init__(self, *args, **kwargs):
...
     """
-    To test postgres-specific general aggregation functions
+    To test postgres-specific general aggregation functions.
```
Comment (timgraham) [inline tests/postgres_tests/models.py]: Try not to make unrelated cosmetic cleanups in a new feature or bug fix.

File: docs/ref/contrib/postgres/aggregates.txt [hunk 3]
```diff
@@ -77,6 +77,15 @@ General-purpose aggregation functions
...
         will be distinct. Defaults to ``False``.
 
+``JsonAgg``
+-----------
+
+.. class:: JsonAgg(expressions, **extra)
+
+    .. versionadded:: 1.11
+
+    Returns the input values concatenated to a ``JSON`` dictionary.
```
Comment (timgraham) [inline docs/ref/contrib/postgres/aggregates.txt]: Is "the input values as a JSON array." a better wording?

File: django/contrib/postgres/aggregates/general.py [hunk 4]
```diff
@@ -42,3 +43,13 @@ def convert_value(self, value, expression, connection, context):
         if not value:
             return ''
         return value
+
+
+class JsonAgg(Aggregate):
```
Comment (timgraham) [inline django/contrib/postgres/aggregates/general.py]: Sort alphabetically?

File: docs/ref/contrib/postgres/aggregates.txt [hunk 5]
```diff
@@ -77,6 +77,15 @@ General-purpose aggregation functions
         An optional boolean argument that determines if concatenated values
         will be distinct. Defaults to ``False``.
 
+``JsonAgg``
```
Comment (timgraham) [inline docs/ref/contrib/postgres/aggregates.txt]: Sort alphabetically?

File: tests/postgres_tests/test_aggregates.py [hunk 6]
```diff
@@ -1,14 +1,22 @@
...
+from django.test.testcases import skipUnlessDBFeature
 from django.test.utils import Approximate
 
 from . import PostgreSQLTestCase
 from .models import AggregateTestModel, StatTestModel
 
+try:
+    from django.contrib.postgres.aggregates import (
+        ArrayAgg, BitAnd, BitOr, BoolAnd, BoolOr, Corr, CovarPop, JsonAgg,
+        RegrAvgX, RegrAvgY, RegrCount, RegrIntercept, RegrR2, RegrSlope, RegrSXX,
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: put `RegrSXX,` on the next line to wrap imports at 79 characters.

File: tests/postgres_tests/test_aggregates.py [hunk 7]
```diff
@@ -110,6 +118,23 @@ def test_string_agg_empty_result(self):
         values = AggregateTestModel.objects.aggregate(stringagg=StringAgg('char_field', delimiter=';'))
         self.assertEqual(values, {'stringagg': ''})
 
+    @skipUnlessDBFeature('has_jsonb_datatype')
+    def test_json_agg(self):
+        """
+        Aggregation for a single value JSON field.
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: I think these docstrings don't add much value. The test method names seem clear.

File: tests/postgres_tests/test_aggregates.py [hunk 8]
```diff
@@ -110,6 +118,23 @@ def test_string_agg_empty_result(self):
...
+        """
+        values = AggregateTestModel.objects.aggregate(jsonagg=JsonAgg('char_field'))
+        self.assertEquals(values, {'jsonagg': ['Foo1', 'Foo2', 'Foo3', 'Foo4']})
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: `assertEqual` the 's' version is a deprecated alias

File: tests/postgres_tests/test_aggregates.py [hunk 9]
```diff
@@ -110,6 +118,23 @@ def test_string_agg_empty_result(self):
...
+
+    @skipUnlessDBFeature('has_jsonb_datatype')
+    def test_empty_json_agg(self):
+        """
+        Test aggregation over an empty queryset.
+        """
+        AggregateTestModel.objects.all().delete()
+        values = AggregateTestModel.objects.aggregate(jsonagg=JsonAgg('integer_field'))
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: You could avoid the delete query with: `AggregateTestModel.objects.none().aggregate(...)`. (I guess the other tests could use the same pattern, not for this PR though.)

File: tests/postgres_tests/test_aggregates.py [hunk 10]
```diff
@@ -110,6 +118,23 @@ def test_string_agg_empty_result(self):
...
+    @skipUnlessDBFeature('has_jsonb_datatype')
+    def test_json_agg(self):
+        """
+        Aggregation for a single value JSON field.
+        """
+        values = AggregateTestModel.objects.aggregate(jsonagg=JsonAgg('char_field'))
+        self.assertEquals(values, {'jsonagg': ['Foo1', 'Foo2', 'Foo3', 'Foo4']})
+
+    @skipUnlessDBFeature('has_jsonb_datatype')
+    def test_empty_json_agg(self):
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: Other tests use naming like: `test_json_agg_empty`

File: tests/postgres_tests/test_aggregates.py [hunk 11]
```diff
@@ -1,14 +1,22 @@
...
 from .models import AggregateTestModel, StatTestModel
 
+try:
+    from django.contrib.postgres.aggregates import (
+        ArrayAgg, BitAnd, BitOr, BoolAnd, BoolOr, Corr, CovarPop, JsonAgg,
+        RegrAvgX, RegrAvgY, RegrCount, RegrIntercept, RegrR2, RegrSlope, RegrSXX,
+        RegrSXY, RegrSYY, StatAggregate, StringAgg,
+    )
+except ImportError:
+    # guarding against test failure on non-postgres environments
```
Comment (timgraham) [inline tests/postgres_tests/test_aggregates.py]: I would say: `pass  # psycopg2 isn't installed`

File: docs/releases/1.11.txt [hunk 12]
```diff
@@ -172,6 +172,9 @@ Minor features
   operation allow using PostgreSQL's ``citext`` extension for case-insensitive
   lookups.
 
+* The new :class:`django.contrib.postgres.aggregates.JsonAgg` allows aggregation
+  of ``JSON`` fields.
```
Comment (timgraham) [inline docs/releases/1.11.txt]: I think "allows aggregating values as a JSON array" is more accurate.

File: tests/postgres_tests/test_aggregates.py (hunk 6 above)
Comment (atombrella) [inline tests/postgres_tests/test_aggregates.py]: Shouldn't this cause the ``isort`` test to fail?
